checker/src/abstract_domains.rs

```diff
@@ -901,7 +908,20 @@ impl AbstractDomain {
             Expression::Mul { left, right } => left.get_as_interval().mul(&right.get_as_interval()),
             Expression::Neg { operand } => operand.get_as_interval().neg(),
             Expression::Sub { left, right } => left.get_as_interval().sub(&right.get_as_interval()),
-            Expression::Variable { .. } => interval_domain::TOP,
+            Expression::Variable { .. } => interval_domain::BOTTOM,
+            Expression::Widened { operand, .. } => {
+                let interval = operand.get_as_interval();
+                if let Expression::Joined { left, .. } = &operand.expression {
+                    let left_interval = left.get_as_interval();
+                    if left_interval.lower_bound() == interval.lower_bound() {
```

Comment:
I may be missing something obvious, but in your description you say:

> When an interval is constructed from a widened join, the leftmost (i.e. earliest) value in the join tree is used to determine if the values in the widened set increase with every iteration, or decrease. If it is the one or the other, the upper (or lower) bound is removed from the returned interval.

But here you are seeing if the lower bounds (or upper bounds) are equal, not increasing/decreasing.

Reply:
The interesting bound is the one that does not change because of the fixed point computation. That one we can keep, the other must be removed. But you are right to question this code. The version you reviewed did not do the right thing when an expression does not have an interval (the bottom case). I have rewritten it and have added comments that hopefully explain things better.